### PR TITLE
split fingerprint + rewrite into two stages to address #118

### PR DESCRIPTION
**Pull request by cbrevdev**
* first stage excludes replaceable source code from fingerprinting
* second stage fingerprints replaceable source code files
  allowing fingerprint changes in non-source code assets to be reflected
  as fingerprint changes in source code assets
* update source code tests assets to have fingerprint match file hash
* extended asset map test to confirm generation is still correct

### CODE REVIEW

**Comment by rickharrison** [comment]
Full disclosure: I am not going to have time to review this in the near future. If someone from the ember team could review, test, and sign off, it would be much appreciated.

**Comment by cbrevdev** [comment]
Understood. I've created an upstream issue to try to gather some interest in reviewing this PR.

To be clear this PR is specific to addressing the case where assets are referenced by source code assets. This side steps the problem of having to consider circular dependencies, at the cost of continuing to miscalculate the fingerprints of source code assets when they import other source code assets.

At the least, it seems like this is the main issue being encountered. As such I'm unsure whether it is worth the effort/complexity of complete asset dependency graph calculation. Especially as the rewrite logic is handled by a separate module (broccoli-asset-rewrite)...

**Comment by makepanic** [comment]
I found a potentially breaking change depending on how the behavior was defined.

I previously added `emojis` inside the `exclude` array. After using the PR dependency, i had to update all exclude values to glob match.

This means something like:
`exclude: ['icons', 'emojis']`

had to be rewritten to:
`exclude: ['**/icons/**', '**/emojis/**]`

I'm not sure if it was correct previously because https://ember-cli.com/user-guide/#fingerprinting-and-cdn-urls says something like `['fonts/169929']` should work

**Comment by mirague** [comment]
Good catch! Perhaps the change to glob syntax could be a PR of its own for the next major release?

**Comment by mirague** [comment]
I wonder who to approach from the Ember team. Perhaps we can ask in #general on the Ember Slack channel.

**Comment by mirague** [comment]
I'm unsure how to get traction for this PR. The repo hasn't been updated this year. Not sure who to approach directly. cc @arjansingh @rwjblue @joliss

File: lib/asset-rev.js [hunk 1]
```diff
@@ -24,9 +25,19 @@ function AssetRev(inputTree, options) {
   this.ignore = options.ignore;
   this.description = options.description;
 
+  // first pass - excludes replaceable source code
+  this.exclude = exclude.slice();
+  this.replaceExtensions.forEach(function(ext) { this.exclude.push('**/*.' + ext); }, this);
```
Comment (rwjblue) [inline lib/asset-rev.js]: Is `this.exclude` a list of glob matchers?  If it is, we should anchor these strings we are pushing to the end of the filename (I wouldn't want to match `assets/.json-generator/something.png` or whatever).

File: lib/asset-rev.js [hunk 2]
```diff
@@ -24,9 +25,19 @@ function AssetRev(inputTree, options) {
...
+  var assetRewrite = UseRev(fingerprintTree, this);
 
-  return UseRev(fingerprintTree, this);
+  // second pass - fingerprints replaceable source code
+  this.exclude = exclude;
+  this.onlyHash = this.replaceExtensions;
+  var fingerprintTree2 = Fingerprint(assetRewrite, this);
+  var assetRewrite2 = UseRev(fingerprintTree2, this);
+
+  return assetRewrite2;
```
Comment (rwjblue) [inline lib/asset-rev.js]: I realize that this isn't new, but I _vaguely_ dislike returning another object from a constructor like this.

File: lib/asset-rev.js (hunk 1 above)
Comment (cbrevdev) [inline lib/asset-rev.js]: `this.exclude` is a list of strings used by minimatch for glob matching, as well as a special-cased direct substring search (indexOf).

    Fingerprint.prototype.excludeContains = function (relativePath) {
      for (var i = 0; i < this.exclude.length; i++) {
        // glob expression
        if (this.exclude[i].match(relativePath)) {
          return true;
        }
        // allow 'fonts' and 'assets/fonts' to match 'assets/font/foo.ttf'
        if (relativePath.indexOf(this.exclude[i].pattern) !== -1) {
          return true;
        }
      }
      return false;
    }

First pass takes advantage of this by excluding the "replaceable source code" file extensions using glob syntax: `js` => `**/*.js`.

Which will match all files ending in the excluded file extension.


I'd also prefer to remove the substring matching altogether, as it can have unexpected behavior as illustrated by your example. But this would be a breaking change...

File: lib/asset-rev.js (hunk 2 above)
Comment (cbrevdev) [inline lib/asset-rev.js]: Yep. I also find this confusing. Can we say this is out of scope of this PR?

**Comment by cbrevdev** [comment]
css `@import` will still be handled incorrectly by this PR. We'd indeed need to change to an dependency graph scheme (mentioned by #29), such as asset-graph. I'd recommend doing this as a separate PR or perhaps a separate Broccoli module?

File: lib/asset-rev.js (hunk 1 above)
Comment (rwjblue) [inline lib/asset-rev.js]: Thank you for the detailed explanation. I'd prefer to avoid the breaking change for now (as you have done here), so that we can fix the primary bug that this is addressing.

File: lib/asset-rev.js (hunk 2 above)
Comment (rwjblue) [inline lib/asset-rev.js]: Yes, it is absolutely out of scope...

**Comment by rwjblue** [comment]
Thank you @cbrevdev!

**Comment by rwjblue** [comment]
Released as v2.6.0

**Comment by rickharrison** [comment]
Thanks @rwjblue !!!

**Comment by rickharrison** [comment]
My app is no longer working. Are we sure this did not introduce a regression?

**Comment by cbrevdev** [comment]
If your app stopped working, then sounds like there could be a regression. While we've moved over to v2.6.0 without issue, and unit tests are passing. It's certainly possible I missed something.

Are you able to provide further details including the fingerprint config?

**Comment by k3n** [comment]
@rickharrison any luck diagnosing the issue?

**Comment by mirague** [comment]
@cbrevdev Do you think this approach could be applied for a similar issue such as https://github.com/kimroen/ember-cli-autoprefixer/issues/38?

**Comment by cbrevdev** [comment]
Thanks for the heads up!

Added comments + PR request over there :)
